### PR TITLE
[C#] Fix interaction with CBv2 and bonded fields 

### PR DESCRIPTION
**Pull request by chadwalters**
Serialization of bonded fields was interacting with the state in the CompactBinaryWriter because it triggered first-pass again. This fix prevents the first-pass writer from being returened in the middle of the
second pass.

### CODE REVIEW

**Comment by chwarr** [comment]
You had mentioned wanting to clean some things up before the rest of us reviewed. Are you ready for review yet?

**Comment by chadwalters** [comment]
Yes.

File: cs/test/core/Util.cs [hunk 1]
```diff
@@ -385,6 +385,18 @@ public static void SerializeCB2<T>(T obj, Stream stream)
             output.Flush();
         }
 
+        public static void SerializeCB2Multiple<T>(T obj, Stream stream, uint times)
+        {
+            var output = new OutputStream(stream, 11);
```
Comment (tstein) [inline cs/test/core/Util.cs]: What's this 11?

File: cs/test/core/BondedTests.cs [hunk 2]
```diff
@@ -438,5 +438,17 @@ public void SerializeBonded()
...
+        [Test]
+        public void SerializeBondedCB2Multiple()
+        {
+            var obj = new StructWithBonded();
+
+            var stream = new MemoryStream();
+            stream.SetLength(0);
+            Util.SerializeCB2Multiple(obj, stream, 2);
+
+            Assert.AreEqual(stream.Position, stream.Position);
```
Comment (tstein) [inline cs/test/core/BondedTests.cs]: This would appear to always succeed.

File: cs/test/core/BondedTests.cs [hunk 3]
```diff
@@ -438,5 +438,17 @@ public void SerializeBonded()
...
         }
+
+        [Test]
+        public void SerializeBondedCB2Multiple()
+        {
+            var obj = new StructWithBonded();
+
+            var stream = new MemoryStream();
+            stream.SetLength(0);
+            Util.SerializeCB2Multiple(obj, stream, 2);
```
Comment (tstein) [inline cs/test/core/BondedTests.cs]: Consider also asserting that the CB writer's state is empty after all that, or - maybe more importantly - that its `.getFirstPassWriter()` returns non-null.

File: cs/test/core/Util.cs (hunk 1 above)
Comment (chadwalters) [inline cs/test/core/Util.cs]: Not entirely sure. Used throughout this file.

File: cs/test/core/BondedTests.cs (hunk 2 above)
Comment (chadwalters) [inline cs/test/core/BondedTests.cs]: Artifact previous revs. Removed.

File: cs/test/core/BondedTests.cs (hunk 3 above)
Comment (chadwalters) [inline cs/test/core/BondedTests.cs]: Added now inside util function (no access to writer here.

File: cs/test/core/BondedTests.cs [hunk 4]
```diff
@@ -438,5 +438,15 @@ public void SerializeBonded()
             Assert.IsTrue(Comparer.Equal(poly1, obj.poly[1].Deserialize()));
             Assert.IsTrue(Comparer.Equal(poly2, obj.poly[2].Deserialize<Derived>()));
         }
+
+        [Test]
```
Comment (chwarr) [inline cs/test/core/BondedTests.cs]: Minor: in the first commit, consider adding `[Ignore("Intentionally failing CBv2 regression test before bug fix")]` to this test, as it's known broken. That way, the build and tests still "pass".

File: cs/src/core/protocols/CompactBinary.cs [hunk 5]
```diff
@@ -176,7 +176,11 @@ public IProtocolWriter GetFirstPassWriter()
         {
             if (version == 2)
             {
-                return firstPassWriter.Value;
+                // Only return first pass if not in middle of second pass
+                if (lengths.Count == 0)
```
Comment (chwarr) [inline cs/src/core/protocols/CompactBinary.cs]: I checked to confirm. `Count` on a `LinkedList` is [constant time](https://msdn.microsoft.com/en-us/library/t56y0h99(v=vs.110).aspx#Anchor_1).

File: cs/test/core/Util.cs [hunk 6]
```diff
@@ -385,6 +385,19 @@ public static void SerializeCB2<T>(T obj, Stream stream)
...
+            var writer = new CompactBinaryWriter<OutputStream>(output, 2);
+
+            for (uint i = 0; i < times; i++)
+            {
+                Serialize.To(writer, obj);
+                Assert.NotNull(writer.GetFirstPassWriter());
```
Comment (chwarr) [inline cs/test/core/Util.cs]: Consider adding a comment about this assertion. Something like:

```csharp
// CBv2 writer should have an active FirstPassWriter before the next top-level serialization.
```

File: cs/test/core/BondedTests.cs (hunk 4 above)
Comment (chadwalters) [inline cs/test/core/BondedTests.cs]: That seems odd to me. Let's discuss general policy offline.

**Comment by chwarr** [comment]
Looks like AppVeyor timed out installing stack.exe. Re-running.

**Comment by chwarr** [comment]
Are you OK if we squash this when merging instead of fixing up the first commit?

**Comment by chadwalters** [comment]
Yes. Go ahead and squash into a single commit.

**Pull request merged**